[VkAlternativeBridge] New bridge

d7sd6u opened the pull request:
Hello. This is an alternative bridge for the vk.com site. I know that a bridge for this site already exists in rss-bridge, but I was dissatisfied with it's code structure, therefore did not want to contribute to it and preferred to write the bridge from scratch. Unlike VkBridge, this bridge correctly embed all types of videos (vk's native videos and videos from Youtube), gifs (inserts the gif itself, not a static preview), images inserted as files (inserts preview images into the feed and adds a direct link to the full-size image ), maps (inserts the map itself and a link to the location, not just the name of the location), posters, files (inserts a direct link to the file, not to the web page with the file), the bridge also allows you to insert comments from the post into feed and subtly set up filtering of comments, plus it is possible to remove from the feed not just posts with reposts, but only posts in which there is nothing but a repost. In addition, the code is structured in a completely different way and separates the parsing of the post from its presentation, allowing to change how to format feed items, for example, this way developer can add css to feed items without touching the parsing at all (as well as the process of getting post data in general). Also, in my opinion, this bridge formats content much more beautifully. The bridge still has some roughness (for example cold start is slow, because of pages downloading), but it is already quite working and suitable for everyday use.

d7sd6u committed:
[VkAlternativeBridge] Alternative bridge for vk.com

d7sd6u committed:
[VkAlternativeBridge] Cleanup spaces

d7sd6u committed:
[VkAlternativeBridge] Fix bugs introduced by refactoring and post plaintext extracting

em92 commented:
Hi, @d7sd6u !
Nice job about rewriting bridge for vk.com.

> I was dissatisfied with it's code structure, therefore did not want to contribute to it and preferred to write the bridge from scratch.

I think you did the right decision about rewriting bridge from scratch. I consider to make it as a replacement, but in this case you would need to add backward compability for old urls to be valid.

Here are some moments:

1. Feed for https://vk.com/a.n.shilov does not work:

https://feed.eugenemolotov.ru/pr1888/?action=display&bridge=VkAlternative&sourceId=a.n.shilov&removeReposts=none&topCommentThreshold=0&branchCommentThreshold=0&descendingCommentThreshold=0&descendingCommentThresholdOffset=0&format=Html

The original vk bridge results: https://feed.eugenemolotov.ru/?action=display&bridge=Vk&u=a.n.shilov&format=Html

2. Also profile name depends on instance country. For example my instance is hosted in Netherlands, so profile is named Alexander Sjilov, while in English it must be Alexander Shilov. Original VkBridge sends some specific headers, to indicate that it requests page in English. Anyway, Russian is also fine, but I remember that it lead to problem, where timestamps were not parsed in original VkBridge.

d7sd6u commented:
Hello, @em92 , thanks for the review. I've been busy this week and will only be able to code again this weekend. On Saturday, Sunday I'll fix the pull request.

> I consider to make it as a replacement, but in this case you would need to add backward compability for old urls to be valid.

I'll try to do backward compatibility, but I need to figure out how to do it for repost filtering (VkBridge uses checkbox, and this bridge uses list).

> Feed for https://vk.com/a.n.shilov does not work:

Most likely this is due to nested reposts (I myself first met them), but I will need to see if this is so. If so, it shouldn't be difficult to fix it.

> Original VkBridge sends some specific headers, to indicate that it requests page in English.

This bridge also sends special headers for this, which, as far as I remember, works in this bridge and VK sends the English version of the page.

d7sd6u committed:
[VkAlternativeBridge] Fixed url compatibility with old VkBridge

d7sd6u committed:
[VkAlternativeBridge] Fixed post extracting for posts with reposted comments

d7sd6u committed:
[VkAlternativeBridge] Fixed feed name extracting

d7sd6u commented:
Good evening, @em92 , I fixed everything, including backward compatibility for old urls.

d7sd6u committed:
[VkAlternativeBridge] Fix feed name locale

d7sd6u committed:
[VkAlternativeBridge] Added post's origin support

d7sd6u committed:
[VkAlternativeBridge] Fixed annoying 'undefined index' warnings

d7sd6u commented:
@em92 , why aren't you merging pull request? In the meantime, I have corrected the feed name extraction (I finally understood what you were talking about in the first comment) and added support for post's origin extracting and formatting.

d7sd6u committed:
[VkAlternativeBridge] Massive pages download optimisation

d7sd6u commented:
Optimized page downloading. Now almost all necessary pages are downloaded in advance and in parallel, using curl_multi. Previously, without caching, feed was created for about a minute or more, but now for three to nine seconds, depending on the number and content of reposts and comments,10-20 times faster! Bridge is now much more usable. In my opinion, such functionality (downloading many links in parallel) should be in rss-bridge libraries, but for this error handling and caching should be added to my function (I will add this later).

d7sd6u committed:
[VkAlternativeBridge] Added error handling and caching to parallel downloading

d7sd6u committed:
[VkAlternativeBridge] Minor optimization

d7sd6u committed:
[VkAlternativeBridge] Fix notice

d7sd6u committed:
[VkAlternativeBridge] Improved compatibility with feed readers

d7sd6u committed:
[VkAlternativeBridge] Added parameter for changing amount of downloaded and parsed posts

d7sd6u committed:
[VkAlternativeBridge] Increased cache timeout

d7sd6u committed:
[VkAlternativeBridge] Fix bug in predownload

d7sd6u committed:
[VkAlternativeBridge] Changed comments separator for better compatibility

d7sd6u commented:
> Optimized page downloading.

However, be careful, in this mode standard 128 megabytes of RAM are not enough for the bridge and therefore I recommend increasing the PHP limit to at least 256 megabytes on your instance.

d7sd6u committed:
[VkAlternativeBridge] Small comments formatting change

d7sd6u commented:
I am rewriting the bridge so that it uses my new libraries, in which I moved all the actions related to parsing and formatting posts, so I close this PR.
